documentation/rir.md

```diff
@@ -0,0 +1,133 @@
+# RIR
+
+## Overview
+
+RIR is a bytecode, similar to assembly or the JVM bytecode. It stands for "R Intermediate Representation".
+
+R is an interpreted language, so every line of code is executed immediately when it's parsed. However, re-parsing every single statement is slow. So, as the RIR interpreter parses code, it first converts it into "RIR bytecode", then executes it. When a call to previously-parsed code is encountered, like in a loop or function call, instead of re-parsing the code, RIR will simply run the existing bytecode. Here is a visualization of how the following might be interpreted (the `>`s denote what the interpreter is reading, note that the actual bytecode generated is more complex):
```

Comment:
The translation from R AST to RIR is a "compilation". The translation from textual source code to R AST would be called parsing. The reason BC is faster to execute, is that it is a denser representation.

Reply:
"similar to assembly or the JVM bytecode" -> I would remove assembly.
"R is an interpreted language" -> I would say that gnu-r is an interpreter not that the language is interpreted.
"so every line of code is executed immediately when it's parsed" -> I guess we parse blocks of code not line by line. Anyway, gnu-r also uses bytecodes nowadays. I would prefer to say here that RIR is another bytecode targeted specifically for the R language and its differences and goals.